sw/device/silicon_creator/lib/rescue/dfu.c

```diff
@@ -0,0 +1,274 @@
+        // FourCC from the value and index fields.
+        case kDfuVendorSetMode: {
+          uint32_t mode = ((uint32_t)setup->value << 16) | setup->index;
+          if (validate_mode(mode, &ctx->state, ctx->bootdata) == kErrorOk) {
+            dfu_transport_data(ctx, kUsbDirIn, NULL, 0, 0);
+          } else {
+            error = kErrorUsbBadSetup;
+          }
+        } break;
+        default:
+          error = kErrorUsbBadSetup;
+      }
+    } else if ((setup->request_type & kUsbReqTypeRecipientMask) ==
+               kUsbReqTypeInterface) {
+      // If its an interface-level request, we handle it here.  These requests
+      // will be {Set,Get}Interface.  We use interface altsettings to control
+      // which services rescue is communicating with.
+      switch (setup->request) {
+        case kUsbSetupReqSetInterface:
+          if (validate_mode(setup->value, &ctx->state, ctx->bootdata) ==
+              kErrorOk) {
+            ctx->interface = (uint8_t)setup->value;
+            dfu_transport_data(ctx, kUsbDirIn, NULL, 0, 0);
+          } else {
+            error = kErrorUsbBadSetup;
+          }
+          break;
+        case kUsbSetupReqGetInterface:
+          dfu_transport_data(ctx, kUsbDirIn, &ctx->interface,
+                             sizeof(ctx->interface), 0);
+          break;
+        default:
+          error = kErrorUsbBadSetup;
+      }
+    } else {
+      // Otherwise, all other requests get mapped to the standard control
+      // endpoint function.
+      error = dfu_transport_setupdata(ctx, setup);
+      // Take care of the SetConfiguration command.
+      if (ctx->ep0.flags & kUsbControlFlagsPendingConfig) {
+        ctx->ep0.flags &= ~(unsigned)kUsbControlFlagsPendingConfig;
+        ctx->ep0.configuration = ctx->ep0.next.configuration;
+        dfu_transport_data(ctx, kUsbDirIn, NULL, 0, 0);
+      }
+    }
+    dfu_transport_result(ctx, error);
+  }
+
+  // If we completed a transaction, process the completion.
+  if (flags & kUsbTransferFlagsDone) {
+    // Take care of the SetAddress command.
+    if (ctx->ep0.flags & kUsbControlFlagsPendingAddress) {
+      ctx->ep0.flags &= ~(unsigned)kUsbControlFlagsPendingAddress;
+      ctx->ep0.device_address = ctx->ep0.next.device_address;
+      usb_set_address(ctx->ep0.device_address);
+    }
+
+    int length = *(int *)data;
+    if (ctx->dfu_state == kDfuStateDnLoadSync) {
+      // If we're in the DnLoadSync state and we completed a transfer, that
+      // means the rescue buffer has data to process.
+      ctx->state.offset = (uint32_t)length;
+      while (ctx->state.offset < sizeof(ctx->state.data)) {
+        // Make sure the full buffer is filled.  Fill unused space with 0xFF.
+        ctx->state.data[ctx->state.offset++] = 0xFF;
+      }
+      // Pass the rescue buffer to the rescue receive handler.
+      rom_error_t error = rescue_recv_handler(&ctx->state, ctx->bootdata);
```

Comment:
Just checking, but the intention is not to support arbitrary transfer sizes for a DFU_DNLOAD request, right?

The DFU protocol allows transfers as small as a single max-sized packet, but it chains multiple transfers to fill up the space. If I'm reading this and `rescue_recv_handler()` correctly, transfers (blocks) that are smaller than a flash page will yield corrupt data, since `flash_firmware_block` will move on to the next page (due to the 0xFF fill here).

Reply:
Correct.  I think technically the spec would allow you to send multiple shorter transfers to fill the buffer before committing to flash.  In practice, `dfu-util` (and opentitantool) send a series of DFU-transfer-sized blocks with the last block possibly being short.

Do you think it is worthwhile to support shorter packets from the host?

Reply:
If you need to violate the spec a bit and require a specific transfer size, that at least won't be a totally new quirk, I think:

https://github.com/fwupd/fwupd/blob/main/plugins/dfu/README.md#dfuforcetransfersize

Reply:
> Correct. I think technically the spec would allow you to send multiple shorter transfers to fill the buffer before committing to flash. In practice, `dfu-util` (and opentitantool) send a series of DFU-transfer-sized blocks with the last block possibly being short.
> 
> Do you think it is worthwhile to support shorter packets from the host?

Hm... If I just think aloud where that _might_ be useful...

Maybe if there is some especially memory-constrained host doing the updating? It seems unlikely that 2 KiB would be a problem for even a microcontroller that has a USB host peripheral, so maybe we're fine. :thinking: 

It'll be good to keep note of our quirks, though.

Reply:
Ok, I've documented the quirks in the implementation, where `kDfuActionDataXfer` is evaluated.